File: tests/test-minecraft-osp/src/main/java/pize/tests/voxelgame/server/player/PlayerList.java
package pize.tests.voxelgame.server.player;

import pize.math.vecmath.vector.Vec3f;
import pize.net.tcp.TcpConnection;
import pize.net.tcp.packet.IPacket;
import pize.tests.voxelgame.main.net.packet.*;
import pize.tests.voxelgame.main.text.Component;
import pize.tests.voxelgame.main.text.TextColor;
import pize.tests.voxelgame.server.Server;
import pize.tests.voxelgame.server.level.ServerLevel;
import pize.tests.voxelgame.server.net.PlayerGameConnection;

import java.util.Collection;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class PlayerList{
    
    private final Server server;
    private final Map<String, ServerPlayer> playerMap;
    
    public PlayerList(Server server){
        this.server = server;
        playerMap = new ConcurrentHashMap<>();
    }
    
    public Server getServer(){
        return server;
    }
    
    
    public Collection<ServerPlayer> getPlayers(){
        return playerMap.values();
    }
    
    public boolean isPlayerOnline(String name){
        return playerMap.containsKey(name);
    }
    
    public ServerPlayer getPlayer(String name){
        return playerMap.get(name);
    }
    
    
    public void addNewPlayer(String name, TcpConnection connection){
        // Get level & Spawn position
        final ServerLevel level;
        final Vec3f spawnPosition;
        
        final OfflinePlayer offlinePlayer = getOfflinePlayer(name);
        if(offlinePlayer != null){
            
            final String levelName = offlinePlayer.getLevelName();
            server.getLevelManager().loadLevel(levelName);
            level = server.getLevelManager().getLevel(levelName);
            spawnPosition = offlinePlayer.getPosition();
        }else{
            
            final String levelName = server.getConfiguration().getDefaultLevelName();
            server.getLevelManager().loadLevel(levelName);
            level = server.getLevelManager().getLevel(levelName);
            spawnPosition = level.getSpawnPosition();
        }
        
        // Add ServerPlayer to list
        final ServerPlayer serverPlayer = new ServerPlayer(level, connection, name);
        server.getConnectionManager().setPacketHandler(connection, serverPlayer.getConnectionAdapter());
        serverPlayer.teleport(level, spawnPosition);
        
        playerMap.put(name, serverPlayer);
        
        // Send packets to player
        final PlayerGameConnection connectionAdapter = serverPlayer.getConnectionAdapter();
        
        new CBPacketSpawnInfo(level.getConfiguration().getName(), spawnPosition).write(connection); // spawn init info
        new CBPacketAbilities(false).write(connection); // abilities
        
        for(ServerPlayer anotherPlayer: playerMap.values())
            if(anotherPlayer != serverPlayer)
                connectionAdapter.sendPacket(new CBPacketSpawnPlayer(anotherPlayer)); // all players info
        
        // Load chunks for player
        level.addEntity(serverPlayer);
        level.getChunkManager().loadInitChunkForPlayer(serverPlayer);

        // Send to all player-connection-event packet
        broadcastPacketExcept(new CBPacketSpawnPlayer(serverPlayer), serverPlayer);
        
        serverPlayer.sendToChat(new Component().color(TextColor.YELLOW).text("Player " + name + " joined the game"));
    }
    
    
    public void disconnectPlayer(ServerPlayer player){
        broadcastPacketExcept(new CBPacketRemoveEntity(player), player); // Remove player entity on client
        player.getLevel().removeEntity(player); // Remove entity on server
        PlayerIO.save(player);                  // Save
        
        playerMap.remove(player.getName());
    }
    
    
    public OfflinePlayer getOfflinePlayer(String name){ //: NICHE TAK, HARAM
        return null;
    }
    
    
    public void broadcastPacket(IPacket<?> packet){
        for(ServerPlayer player: playerMap.values())
            player.sendPacket(packet);
    }

    public void broadcastPacketExcept(IPacket<?> packet, ServerPlayer except){
        for(ServerPlayer player: playerMap.values())
            if(player != except)
                player.sendPacket(packet);
    }

    public void broadcastPacketLevel(ServerLevel level, IPacket<?> packet){
        for(ServerPlayer player: playerMap.values())
            if(player.getLevel() == level)
                player.sendPacket(packet);
    }

    public void broadcastPacketLevelExcept(ServerLevel level, IPacket<?> packet, ServerPlayer except){
        for(ServerPlayer player: playerMap.values())
            if(player.getLevel() == level && player != except)
                player.sendPacket(packet);
    }
    
    
    public void broadcastServerMessage(Component message){
        broadcastPacket(new CBPacketChatMessage(message.toFlatList()));
        System.out.println(message);
    }
    
    public void broadcastPlayerMessage(ServerPlayer player, Component message){
        broadcastPacket(new CBPacketChatMessage(player.getName(), message.toFlatList()));
        System.out.println(message);
    }
    
}

File: tests/test-minecraft-osp/src/main/kotlin/pize/tests/voxelgame/server/player/PlayerList.kt
package pize.tests.voxelgame.server.player

import pize.math.vecmath.vector.Vec3f
import pize.net.tcp.TcpConnection
import pize.net.tcp.packet.IPacket
import pize.tests.voxelgame.main.net.packet.*
import pize.tests.voxelgame.main.text.Component
import pize.tests.voxelgame.main.text.TextColor
import pize.tests.voxelgame.server.Server
import pize.tests.voxelgame.server.level.ServerLevel
import java.util.concurrent.ConcurrentHashMap

class PlayerList(val server: Server) {
    private val playerMap: MutableMap<String?, ServerPlayer>

    init {
        playerMap = ConcurrentHashMap()
    }

    val players: Collection<ServerPlayer>
        get() = playerMap.values

    fun isPlayerOnline(name: String?): Boolean {
        return playerMap.containsKey(name)
    }

    fun getPlayer(name: String?): ServerPlayer? {
        return playerMap[name]
    }

    fun addNewPlayer(name: String?, connection: TcpConnection?) {
        // Get level & Spawn position
        val level: ServerLevel?
        val spawnPosition: Vec3f?
        val offlinePlayer = getOfflinePlayer(name)
        if (offlinePlayer != null) {
            val levelName = offlinePlayer.levelName
            server.levelManager.loadLevel(levelName)
            level = server.levelManager.getLevel(levelName)
            spawnPosition = offlinePlayer.position
        } else {
            val levelName = server.configuration.defaultLevelName
            server.levelManager.loadLevel(levelName)
            level = server.levelManager.getLevel(levelName)
            spawnPosition = level.getSpawnPosition()
        }

        // Add ServerPlayer to list
        val serverPlayer = ServerPlayer(level, connection, name)
        server.connectionManager.setPacketHandler(connection, serverPlayer.connectionAdapter)
        serverPlayer.teleport(level, spawnPosition)
        playerMap[name] = serverPlayer

        // Send packets to player
        val connectionAdapter = serverPlayer.connectionAdapter
        CBPacketSpawnInfo(level.configuration.name, spawnPosition).write(connection) // spawn init info
        CBPacketAbilities(false).write(connection) // abilities
        for (anotherPlayer in playerMap.values) if (anotherPlayer !== serverPlayer) connectionAdapter!!.sendPacket(
            CBPacketSpawnPlayer(anotherPlayer)
        ) // all players info

        // Load chunks for player
        level.addEntity(serverPlayer)
        level.chunkManager.loadInitChunkForPlayer(serverPlayer)

        // Send to all player-connection-event packet
        broadcastPacketExcept(CBPacketSpawnPlayer(serverPlayer), serverPlayer)
        serverPlayer.sendToChat(Component().color(TextColor.YELLOW).text("Player $name joined the game"))
    }

    fun disconnectPlayer(player: ServerPlayer?) {
        broadcastPacketExcept(CBPacketRemoveEntity(player), player) // Remove player entity on client
        player.getLevel().removeEntity(player) // Remove entity on server
        PlayerIO.save(player) // Save
        playerMap.remove(player.getName())
    }

    fun getOfflinePlayer(name: String?): OfflinePlayer? { //: NICHE TAK, HARAM
        return null
    }

    fun broadcastPacket(packet: IPacket<*>) {
        for (player in playerMap.values) player.sendPacket(packet)
    }

    fun broadcastPacketExcept(packet: IPacket<*>, except: ServerPlayer?) {
        for (player in playerMap.values) if (player !== except) player.sendPacket(packet)
    }

    fun broadcastPacketLevel(level: ServerLevel, packet: IPacket<*>) {
        for (player in playerMap.values) if (player.level === level) player.sendPacket(packet)
    }

    fun broadcastPacketLevelExcept(level: ServerLevel?, packet: IPacket<*>, except: ServerPlayer) {
        for (player in playerMap.values) if (player.level === level && player !== except) player.sendPacket(packet)
    }

    fun broadcastServerMessage(message: Component?) {
        broadcastPacket(CBPacketChatMessage(message!!.toFlatList()))
        println(message)
    }

    fun broadcastPlayerMessage(player: ServerPlayer, message: Component?) {
        broadcastPacket(CBPacketChatMessage(player.name, message!!.toFlatList()))
        println(message)
    }
}
